src/bika/lims/content/client.py

```diff
@@ -173,29 +173,118 @@ class Client(Organisation):
 
     security = ClassSecurityInfo()
-    displayContentsTab = False
     schema = schema
-    _at_rename_after_creation = True
+    GROUP_KEY = "_client_group_id"
 
     def _renameAfterCreation(self, check_auto_id=False):
         from bika.lims.idserver import renameAfterCreation
         renameAfterCreation(self)
 
-    security.declarePublic("getContactFromUsername")
+    @property
+    def group_id(self):
+        """Client group ID
+        """
+        if not hasattr(self, self.GROUP_KEY):
+            setattr(self, self.GROUP_KEY, self.getClientID() or self.getId())
+        return getattr(self, self.GROUP_KEY)
+
+    @security.public
+    def get_group(self):
+        """Returns our client group
+
+        :returns: Client group object
+        """
+        portal_groups = api.get_tool("portal_groups")
+        return portal_groups.getGroupById(self.group_id)
+
+    @security.private
+    def create_group(self):
+        """Create a new client group
+
+        :returns: Client group object
+        """
+        group = self.get_group()
+
+        # return the existing Group immediately
+        if group:
+            return group
+
+        portal_groups = api.get_tool("portal_groups")
+        group_name = self.getName()
+
+        # Create a new Client Group
+        # NOTE: The global "Client" role is necessary for the client contacts
+        portal_groups.addGroup(self.group_id,
+                               roles=["Client"],
+                               title=group_name)
 
+        # Grant the group the "Owner" role on ourself
+        # NOTE: This will grant each member of this group later immediate
+        #       access to all exisiting objects with the same role.
+        self.manage_setLocalRoles(self.group_id, ["Owner"])
+
+        return self.get_group()
+
+    @security.private
+    def remove_group(self):
+        """Remove the own client group
+
+        :returns: True if the group was removed, otherwise False
+        """
+        group = self.get_group()
+        if not group:
+            return False
+        portal_groups = api.get_tool("portal_groups")
+        # Use the client ID for the group ID
+        group_id = self.getClientID()
+        portal_groups.removeGroup(group_id)
+        return True
+
+    @security.private
+    def add_user_to_group(self, user):
+        """Add a user to the client group
+
+        :param user: user/group object or user/group ID
+        :returns: list of new group IDs
+        """
+        group = self.get_group()
+        if not group:
+            group = self.create_group()
+        return api.user.add_group(group, user)
+
+    @security.private
+    def del_user_from_group(self, user):
+        """Add a user to the client group
+
+        :param user: user/group object or user/group ID
+        :returns: list of new group IDs
+        """
+        group = self.get_group()
+        if not group:
+            group = self.create_group()
```

Comment:
Is this necessary? Why to not return `False` directly if there is no group?

Reply:
Basically I want to ensure a group if it was either deleted, or not created due to the registry setting. Remember, the group has the global `Client` role and the local `Owner` role on the current client object.